<a href="https://colab.research.google.com/github/rtajeong/AI_Cluster_MSAC/blob/main/gg_95_OpenCV_%EB%8F%99%EC%98%81%EC%83%81%EC%B2%98%EB%A6%AC_rev2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [17]:
# !pip install pytube
!pip install pytubefix

In [18]:
import cv2
from google.colab import drive
# from pytube import YouTube
from pytubefix import YouTube

In [32]:
# Youtube 객체 생성 (제목, 길이, 조회수, 스트림 정보 등 메타 데이터 포함)
yt = YouTube("https://www.youtube.com/watch?v=OwJPPaEyqhI")

# 스트림 필터링 (영상과 오디오가 합쳐진 스트림만 선택, 가장 높은 해상도)
yt = yt.streams.filter(progressive=True, file_extension='mp4').order_by('resolution').desc().first()
yt.download('/content/')

'/content/PSY - ‘New Face’ MV.mp4'

In [20]:
# 비디오 경로 지정
# file name 을 못 읽으면 New_Face.mp4 로 수정할 것
video_path = "/content/New_Face.mp4"

In [21]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [28]:
video = cv2.VideoCapture(video_path) # 비디오 파일 열기
fps = video.get(cv2.CAP_PROP_FPS)    # 비디오의 초당 프레임 수(FPS, Frames Per Second)를 가져오는 함수
fps

23.976023976023978

In [30]:
frame_count = int(video.get(cv2.CAP_PROP_FRAME_COUNT))
print("총 프레임 수:", frame_count)
duration = frame_count / fps                     # 영상 길이(초)
print("FPS:", fps)
print("총 프레임 수:", frame_count)
print("영상 길이(초):", duration)

총 프레임 수: 4825
FPS: 23.976023976023978
총 프레임 수: 4825
영상 길이(초): 201.24270833333333


In [23]:
import os
video = cv2.VideoCapture(video_path)
success = True
count = 1

while success:
  success,frame = video.read()
  if not os.path.exists("/content/test"):
      os.mkdir('test')
  name = '/content/test/'+str(count)+'.jpg'

  if success == True:
    cv2.imwrite(name, frame)
    print('Frame {} Extracted Successfully'.format(count))
    count = count+100       # 100 프레임마다 한 개씩
    if (count == 10000):    # just for the first 10 frames
        break
  else:
    print("failed to read YouTube video.")
    break

Frame 1 Extracted Successfully
Frame 101 Extracted Successfully
Frame 201 Extracted Successfully
Frame 301 Extracted Successfully
Frame 401 Extracted Successfully
Frame 501 Extracted Successfully
Frame 601 Extracted Successfully
Frame 701 Extracted Successfully
Frame 801 Extracted Successfully
Frame 901 Extracted Successfully
Frame 1001 Extracted Successfully
Frame 1101 Extracted Successfully
Frame 1201 Extracted Successfully
Frame 1301 Extracted Successfully
Frame 1401 Extracted Successfully
Frame 1501 Extracted Successfully
Frame 1601 Extracted Successfully
Frame 1701 Extracted Successfully
Frame 1801 Extracted Successfully
Frame 1901 Extracted Successfully
Frame 2001 Extracted Successfully
Frame 2101 Extracted Successfully
Frame 2201 Extracted Successfully
Frame 2301 Extracted Successfully
Frame 2401 Extracted Successfully
Frame 2501 Extracted Successfully
Frame 2601 Extracted Successfully
Frame 2701 Extracted Successfully
Frame 2801 Extracted Successfully
Frame 2901 Extracted Succe

# For reference

- YouTube object

In [4]:
yt = YouTube("https://youtu.be/OwJPPaEyqhI")

In [5]:
yt.title

'PSY - ‘New Face’ M/V'

In [6]:
yt.views

311609635

In [ ]:
# yt.vid_info

In [8]:
# dir(yt)

In [9]:
for e in yt.streams:
    print(e)

<Stream: itag="18" mime_type="video/mp4" res="360p" fps="24fps" vcodec="avc1.42001E" acodec="mp4a.40.2" progressive="True" sabr="False" type="video">
<Stream: itag="137" mime_type="video/mp4" res="1080p" fps="24fps" vcodec="avc1.640028" progressive="False" sabr="False" type="video">
<Stream: itag="248" mime_type="video/webm" res="1080p" fps="24fps" vcodec="vp9" progressive="False" sabr="False" type="video">
<Stream: itag="399" mime_type="video/mp4" res="1080p" fps="24fps" vcodec="av01.0.08M.08" progressive="False" sabr="False" type="video">
<Stream: itag="136" mime_type="video/mp4" res="720p" fps="24fps" vcodec="avc1.4d401f" progressive="False" sabr="False" type="video">
<Stream: itag="247" mime_type="video/webm" res="720p" fps="24fps" vcodec="vp9" progressive="False" sabr="False" type="video">
<Stream: itag="398" mime_type="video/mp4" res="720p" fps="24fps" vcodec="av01.0.05M.08" progressive="False" sabr="False" type="video">
<Stream: itag="135" mime_type="video/mp4" res="480p" fps="2

- 유튜브에서는 음성과 영상을 따로 분리된 채로 가지고 있다가, 스트리밍할 때에 이 두 가지를 하나로 합쳐서 서비스한다.
- 유튜브의 영상은 하나의 영상에 대해서 여러가지 화질과 여러가지 확장자의 버전을 제공한다.(전파 환경에 따라 자동으로 화질을 변경할 수 있도록 - DASH(dynamic adaptdive dtreaming over HTTP))
- 위 Stream 중에서 mp4 확장자만 따로 추출하고 싶으면 filter()를 사용한다.
- video 와 audio 가 있다. (audio 는 영상 없이 음성만 담겨 있음)
- video 중에서는 audio_codec 이 없는 것도 있음. (영상만 있음)
- DASH 관련 Stream을 제거하려면 Progressive = True (progressive mode 에서의 최대 해상도는 720 p)

In [10]:
for e in yt.streams.filter(file_extension='mp4'):
    print(e)

<Stream: itag="18" mime_type="video/mp4" res="360p" fps="24fps" vcodec="avc1.42001E" acodec="mp4a.40.2" progressive="True" sabr="False" type="video">
<Stream: itag="137" mime_type="video/mp4" res="1080p" fps="24fps" vcodec="avc1.640028" progressive="False" sabr="False" type="video">
<Stream: itag="399" mime_type="video/mp4" res="1080p" fps="24fps" vcodec="av01.0.08M.08" progressive="False" sabr="False" type="video">
<Stream: itag="136" mime_type="video/mp4" res="720p" fps="24fps" vcodec="avc1.4d401f" progressive="False" sabr="False" type="video">
<Stream: itag="398" mime_type="video/mp4" res="720p" fps="24fps" vcodec="av01.0.05M.08" progressive="False" sabr="False" type="video">
<Stream: itag="135" mime_type="video/mp4" res="480p" fps="24fps" vcodec="avc1.4d401e" progressive="False" sabr="False" type="video">
<Stream: itag="397" mime_type="video/mp4" res="480p" fps="24fps" vcodec="av01.0.04M.08" progressive="False" sabr="False" type="video">
<Stream: itag="134" mime_type="video/mp4" re

In [11]:
yt.streams.filter(file_extension='mp4').order_by('resolution')

[<Stream: itag="160" mime_type="video/mp4" res="144p" fps="24fps" vcodec="avc1.4d400c" progressive="False" sabr="False" type="video">, <Stream: itag="394" mime_type="video/mp4" res="144p" fps="24fps" vcodec="av01.0.00M.08" progressive="False" sabr="False" type="video">, <Stream: itag="133" mime_type="video/mp4" res="240p" fps="24fps" vcodec="avc1.4d4015" progressive="False" sabr="False" type="video">, <Stream: itag="395" mime_type="video/mp4" res="240p" fps="24fps" vcodec="av01.0.00M.08" progressive="False" sabr="False" type="video">, <Stream: itag="18" mime_type="video/mp4" res="360p" fps="24fps" vcodec="avc1.42001E" acodec="mp4a.40.2" progressive="True" sabr="False" type="video">, <Stream: itag="134" mime_type="video/mp4" res="360p" fps="24fps" vcodec="avc1.4d401e" progressive="False" sabr="False" type="video">, <Stream: itag="396" mime_type="video/mp4" res="360p" fps="24fps" vcodec="av01.0.01M.08" progressive="False" sabr="False" type="video">, <Stream: itag="135" mime_type="video/m

In [12]:
yt.streams.filter(progressive=True,file_extension='mp4').order_by('resolution').desc()

[<Stream: itag="18" mime_type="video/mp4" res="360p" fps="24fps" vcodec="avc1.42001E" acodec="mp4a.40.2" progressive="True" sabr="False" type="video">]

In [13]:
yt.streams.filter(file_extension='mp4').order_by('resolution').desc().first()

<Stream: itag="399" mime_type="video/mp4" res="1080p" fps="24fps" vcodec="av01.0.08M.08" progressive="False" sabr="False" type="video">

In [14]:
yt.streams.filter(progressive=True, file_extension='mp4').order_by('resolution').desc().first()

<Stream: itag="18" mime_type="video/mp4" res="360p" fps="24fps" vcodec="avc1.42001E" acodec="mp4a.40.2" progressive="True" sabr="False" type="video">

# Exercise

In [15]:
!pip install pytube
from pytube import YouTube
import cv2

# 유튜브 영상 다운로드
url = "https://www.youtube.com/watch?v=wPsVStI1M4w"
yt = YouTube(url)
stream = yt.streams.filter(file_extension='mp4').first()
stream.download(filename='video.mp4')

# OpenCV로 프레임 읽기
cap = cv2.VideoCapture('video.mp4')
ret, frame = cap.read()

# 첫 프레임 저장
if ret:
    cv2.imwrite('frame0.jpg', frame)

cap.release()


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 57.6/57.6 kB 1.8 MB/s eta 0:00:00


HTTPError: HTTP Error 400: Bad Request